### ISSUE TITLE
Perimeta Rejects Notify from Sprout with an Apparent Invalid AOR 

### ISSUE THREAD

**Issue by dcollins991** [Metaswitch/sprout#1274]
For the "REGISTER - Network Dereg" test scenario, a UE registers successfully(300 expiry) and then SUBSCRIBES to registration events.  The UE receives a NOIFY and then sends a 200 OK.  The UE then waits 5 mutes and waits on a NOTIFY.  After 5 minutes Sprout issues a NOTIFY and Perimeta rejects with a 403 stating, 
"A call to an unknown subscriber was rejected because the adjacency does 
not permit such calls. Check the public ID and local ID to see if an 
attempt to register has been rejected. 
Public ID = sip:7030000010@mvoip.richlab.datcon.co.uk 
Local ID = sip:quaff@192.168.162.158:32144;transport=udp;ob"
Not sure if this is the root cause of the issue but the aor is invalid and appears as, 
registration aor="aor" id="209243057466376192" state="active"  or, 
registration aor="aor" id="209243057466376192" state="terminated"
Sprout earlier reported an AOR in the SAS log of, 
RegStore SET for AoR: "sip:7030000010@mvoip.richlab.datcon.co.uk" succeeded.  
None of the AOR values in the Notify xml are properly set.  In Release Rob Roy it was set in the register  Notify but not in the dereg NOTIFY.

I have SAS traces and diags.  Please clarify as to whether Sprout is expected to support the registration aor field.

**Comment by mirw** [comment]
@eleanor-merry, I think you've done some work in this area - please can you take a look?

**Comment by rkd-msw** [comment]
I've taken a look at this.

* The AOR should indeed be "sip:7030000010@mvoip.richlab.datcon.co.uk" rather than "aor", and this looks like a small fix.
* However, I think this is a longstanding bug and not a regression (though the issue suggests otherwise, at least in some cases) - so I don't think this is the root cause. I wonder if in fact the problem is that we're trying to send a NOTIFY to a UE saying that that UE's registration has expired (which is obviously doomed).

@dcollins991 - I can't see an issue1274 folder with diags in the usual place. can you upload them? If possible, can you include a trace from the Rob Roy release, so I can diagnose the differences you're reporting in that release?

Can you also confirm if you have registration monitoring on in the P-CSCF or not?

**Comment by dcollins991** [comment]
The diags and SAS traces are in the issue1274 folder.  Sorry, no longer have the Rob Roy trace.It was similar to the Slachthof-Funf load except the initial Notify had the aor set properly.  No I did not have the registration monitoring turned on.  I turned it on and got the same results.  The SAS trace for registration monitoring is also there (see NetworkDeregwRegMonitor.sas.  You can access our sas for vm3 on,http://sas-1.vm3.richlab.datcon.co.uk/serviceassurance .

**Comment by rkd-msw** [comment]
Yep, we're sending the registration to the UE whose registration has expired:

```
NOTIFY sip:quaff@192.168.162.158:30794;transport=tcp;ob SIP/2.0
...
Subscription-State: terminated;reason=deactivated
...

<?xml version="1.0" encoding="UTF-8"?>
<reginfo xmlns="urn:ietf:params:xml:ns:reginfo" xmlns:gr="urn:ietf:params:xml:ns:gruuinfo" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="0" state="full">
<registration aor="aor" id="208839458315829248" state="terminated">
  <contact id="&lt;urn:uuid:f2f42cc6-7f0e-5713-8b54-56347478d6b2&gt;" state="terminated" event="expired">
   <uri>sip:quaff@192.168.162.158:30794;transport=TCP;ob</uri>
   <gr:pub-gruu uri="sip:7030000010@mvoip.richlab.datcon.co.uk;gr=urn:uuid:f2f42cc6-7f0e-5713-8b54-56347478d6b2" />
  </contact>
</registration>
</reginfo>
```

(Note that the Req-URI and the URI element of the expired binding are the same.)

Fix is probably to fix up `SubscriberDataManager::expire_subscriptions` to expire subscriptions which originated from a binding which has now expired.

**Comment by dcollins991** [comment]
Just to clarify from RFC 3083,
"It is RECOMMENDED that a user be allowed to subscribe to their own
   registration state.  Such subscriptions are useful when there are
   many devices that represent a user, each of which needs to learn the
   registration state of the other devices.  We also anticipate that
   applications and automata will frequently be subscribers to the
   registration state.  In those cases, authorization policy will
   typically be provided ahead of time."
We aren't going to support this so I shouldn't test for a user to be able to receive their on dereg events, right?  (Picked up from commit comment: "When binding A expires, don't send a NOTIFY to binding A announcing the subscription is terminated - that's pointless")
The test scenario also poses an issue for Perimeta and TCP.  Its a race between Perimeta shutting down the interface and Perimeta being able to send the Notify out for the dereg event.

**Comment by rkd-msw** [comment]
Hi David,

We support users subscribing to their own registration state in general. This specific issue is about the NOTIFY saying that binding A has been deregistered - as binding A has now gone, then even if there was a subscription over binding A, it doesn't make sense to send a NOTIFY to that (now-expired) binding, as the P-CSCF won't route it and the UE may already have shut down.

However, it is valid to test that a **user** can receive their own dereg events. One way you might test this is:

* user sip:bob@example.com registers UE 1 and UE 2
* from UE 2, they subscribe to their own registration state
* UE 1 deregisters
* UE 2 should get a NOTIFY saying that UE 1 has deregistered

Does that make sense?

**Comment by dcollins991** [comment]
Good!  It avoids the race condition.  I had already started on doing what you suggested.  Thanks!

**Comment by rkd-msw** [comment]
https://github.com/Metaswitch/sprout/pull/1275 (which will be in release-89) should have fixed the issues here - we'll use the real AOR instead of the hardcoded string "aor", and we won't send NOTIFYs to bindings that we know are expired.

Can you close this issue once you're happy with the behaviour?

**Comment by sathiyan-sivathas** [comment]
Fixed by #1275 and UTs test all the scenarios so I'll close this.
